src/rrule.ts

```diff
@@ -234,11 +238,51 @@ export default class RRule implements QueryMethods {
     return result as Date
   }
 
+  /**
+   * Returns the very first occurrence of this rrule if it exists, null otherwise
+   *
+   * @returns {Date | null}
+   */
+  first (): Date | null {
+    return this.after(this.options.dtstart, true) || null
+  }
+
+  /**
+   * Returns the last occurrence of this rrule if it exists, null otherwise
+   * @returns {Date | null}
+   */
+  last (): Date | null {
+    if (!this.isFinite()) {
+      return null
+    }
+
+    if (this.options.until) {
+      return this.before(this.options.until, true)
+    }
+
+    const allDates = this.all()
```

Comment:
Suggestion here would be to use this.before() with a max date far in the future (year 9999, etc.) to avoid having all() save every date to memory while calculating the last date.